lib/vrl/stdlib/src/reverse.rs

```diff
@@ -0,0 +1,78 @@
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Reverse;
+
+impl Function for Reverse {
+    fn identifier(&self) -> &'static str {
+        "reverse"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::ARRAY,
+            required: true,
+        }]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "reverse array",
+            source: r#"reverse([0, 1])"#,
+            result: Ok("[1, 0]"),
+        }]
+    }
+
+    fn compile(&self, _state: &state::Compiler, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+
+        Ok(Box::new(ReverseFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ReverseFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ReverseFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let mut value = self.value.resolve(ctx)?.try_array()?;
+
+        value.reverse();
+
+        Ok(value.into())
+    }
+
+    fn type_def(&self, _state: &state::Compiler) -> TypeDef {
+        TypeDef::new().array_mapped::<(), Kind>(map! { (): Kind::all() })
```

Comment:
This could be improved, although admittedly through a slight rabbit hole of complexity.

Currently this is saying that the return is an array, but we don't know what the types of the array elements are.

There is one situation under which we do know what all the types are. If the parameter array is of type `Any -> x`, where x is some known type (`KindInfo`). For example, if the parameter is an array of integers - we know the result will also be an array of integers.

If you are willing to take on this challenge, you will need to add a function to `lib/vrl/compiler/src/type_def.rs`, perhaps also called `reverse`.  Have a look at the `restrict_array` function. `reverse` would be similar to this, but it would need to check that the restricted array only has a single element in its `BTreeMap`, and the `Index` would have to be `Index::Any`. If it does, we can return that array as the typedef.

If it doesn't, we can't guarantee what the types of the reversed array would be, so we should just return `Kind::all()` as the current implementation does.

The reason we can't guarantee the types for anything other than `Index::Any` is because the array can come from the  event passed into VRL. We don't know what the type of this array is, or how many elements it has. One could theoretically set a particular index in this array, which would result in the typedef for that array knowing the type for that given index. However, since we don't know the length of that array, when reversing we have no way of knowing where the value at that index will end up.

The return value could then be something like:

```rust
    fn type_def(&self, state: &state::Compiler) -> TypeDef {
        self.value.type_def(state).reverse()
    }
```

I hope that makes sense. The typedefs can get a bit hairy because we are trying to impose a type system on top of untyped values being passed in, so the type system has to cater for the unknowns as well as the knowns. Feel free to fire over any questions if you need!